lib/compiler.js

```diff
@@ -461,25 +466,38 @@ function buildMiddlewareInstructions(rootDir, config) {
         allConfigs = [allConfigs];
 
       allConfigs.forEach(function(config) {
-        var resolved = resolveMiddlewarePath(rootDir, middleware);
-
-        var middlewareConfig = cloneDeep(config);
-        middlewareConfig.phase = phase;
 
-        if (middlewareConfig.params) {
-          middlewareConfig.params = resolveMiddlewareParams(
-            rootDir, middlewareConfig.params);
+        var resolved = resolveMiddlewarePath(rootDir, middleware, config);
+
+        /*
+        * resolved.sourceFile will be `false` if an optional middleware
+        * is not resolvable.
+        * if a non-optional middleware is not resolvable, it will throw
+        * at resolveAppPath() and not reach here
+        */
+        if (resolved.sourceFile) {
+          var middlewareConfig = cloneDeep(config);
+          middlewareConfig.phase = phase;
+
+          if (middlewareConfig.params) {
+            middlewareConfig.params = resolveMiddlewareParams(
+              rootDir, middlewareConfig.params);
+          }
+
+          var item = {
+            sourceFile: resolved.sourceFile,
+            config: middlewareConfig
+          };
+          if (resolved.fragment) {
+            item.fragment = resolved.fragment;
+          }
+          middlewareList.push(item);
+        } else {
+          debug('Middleware "%s" not found: %s', middleware, resolved.optional);
```

Comment:
Please use a reverse condition and an early return, that will reduce the amount of indentation changes in the diff. Also use `console.log` instead of `debug` to print the message.

```js
if (!resolved.sourceFile) {
  // print message;
  return; // from forEach()
}

// proceed with the original code
```